File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.cs
using System;
using System.Collections.Generic;

public class Solution {
    private int[] dx = {1, -1, 0, 0};
    private int[] dy = {0, 0, 1, -1};
    public int solution(string[] maps) {
        
        int n = maps.Length;
        int m = maps[0].Length;
                
        int sX = 0; int sY = 0;
        int lX = 0; int lY = 0;
               
        for(int i = 0; i < n; i++)
        {
            for(int j = 0;  j < m; j++)
            {
                char c = maps[i][j];
                if(c == 'S') { sX = i; sY = j; }
                else if(c == 'L') { lX = i; lY = j; }
            }
        }
        
        int toLever = BFS(maps, sX, sY, 'L');
        if(toLever == -1) return -1;
        
        int toExit = BFS(maps, lX, lY, 'E');
        if(toExit == -1) return -1;
        
        return toLever + toExit;
    }
    
    private int BFS(string[] maps, int startX, int startY, char target)
    {
        int n = maps.Length;
        int m = maps[0].Length;
        bool[,] visited = new bool[n, m];
    
        Queue<(int, int, int)> q = new Queue<(int, int, int)>();
        q.Enqueue((startX, startY, 0));
        visited[startX, startY] = true;
        
        while(q.Count > 0)
        {
            var (x, y, d) = q.Dequeue();
            for(int i = 0; i < 4; i++)
            {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                
                char cell = maps[nx][ny];
                if(cell == 'X') continue;
                if(visited[nx, ny]) continue;
                if(cell == target) return d + 1;
                
                visited[nx, ny] = true;
                q.Enqueue((nx, ny, d + 1));
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};
int visited[101][101];

int bfs(int sy, int sx, int ey, int ex, vector<string> maps)
{
    memset(visited, 0, sizeof(visited));
    
    queue<pair<int, int>> q;
    q.push({sy, sx});
    visited[sy][sx] = 1;
    
    while(q.size())
    {
        auto cur = q.front();
        q.pop();
        
        int y = cur.first;
        int x = cur.second;
        
        if(y == ey && x == ex)
            return visited[y][x] - 1;
        
        for(int i = 0; i < 4; i++)
        {
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if(ny < 0 || nx < 0 || ny >= maps.size() || nx >= maps[0].size())
                continue;
            
            if(visited[ny][nx] || maps[ny][nx] == 'X')
                continue;
            
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny, nx});
        }
    }
    return -1;
}

int solution(vector<string> maps) {
    int sy, sx, ly, lx, ey, ex;
    
    for(int i = 0; i < maps.size(); i++)
    {
        for(int j = 0; j < maps[i].size(); j++)
        {
            if(maps[i][j] == 'S') sy = i, sx = j;
            else if(maps[i][j] == 'L') ly = i, lx = j;
            else if(maps[i][j] == 'E') ey = i, ex = j;
        }
    }
    
    int a = bfs(sy, sx, ly, lx, maps);
    if(a == -1) return -1;
    
    int b = bfs(ly, lx, ey, ex, maps);
    if(b == -1) return -1;
    
    return a + b;
}

// 먼저 start 지점에서 레버까지 최단거리로 이동후, exit로 최단거리로 이동
// 레버까지 이동후 visited 배열 초기화 필요
// bfs 사용(queue)